### DISCUSSION TITLE
Qwerty-mixed Plover?

### DISCUSSION THREAD

**Discussion by user202729** [opensteno/plover#1283] [Ideas]
To partially eliminate the difficulty people get when getting started with steno, I think it's feasible to have some system to detect if a key press is "normal Qwerty" or "steno", so people can start using steno without spend too much effort figuring out chords for basic words.

It's still required to learn the theory, although that's only a few pages and only take about an hour of training.  
The hard part is the "unwritten" details of the theory (mostly "how to split words")

Perhaps adapt the suggestions system to suggest strokes for words that the user typed in Qwerty too.

More ideas:

* Suggestion should be in pseudo-steno
* use Spectra Lexer for visualization/other nice things)
* Only suggest words user typed in Qwerty

Unlike normal Plover usage (where it's possible to arpeggiate), it's required to have a NKRO keyboard.

**Comment by panathea** [comment]
I think writing such an algorithm wouldn't be too hard… maybe:

- Of course, if any nonsteno keys are pressed, don't try to detect chords.
- If keys in a chord are all pressed down at the same time, then it's a chord. This is as opposed to the rolling-style of fast typing, like "t down, h, down, t up, e down, h up, e up", contrasted with t, h, and e down; which would be `P*` with two asterisks held down 😆
- We'd probably want a way to handle single-key chords, and I'd propose a hold-down time like 300ms? This would interrupt key-repeat, but I think running Plover in the background would interrupt key-repeat anyway.


As an aside, the "unwritten" details of the theory (how to split words) is something I plan on writing about.

**Reply by user202729** [reply]
Anyone want to write a visualization program to see what's the gap between key presses in chord/typing?

Okay (as expected. It took me a long time to get to a state that I can write most of English with Plover without looking up)

There are only 23 single-key chord anyway, and within those:

* Used: `is it can with had are of be {^'ll} {^ing} the {^s} {^ed} {a^} he you`
* Duplicate: `are <s> (R-/-R, -S/-Z)`
* Misstroke: `about P-, to O-, {.} -P`
* Not used: `#`

Changing these 16 can't be too hard. Besides the words most of the time form phrases.

`-P` and `-L` are sometimes used as suffix,
but they can be changed to `*P` and `*L` respectively.

**Reply by panathea** [reply]
*(Moved comment as replying by email made it a new response)*

I think we can't assume people use english steno and the plover dictionary
for this use case. Single key chords should be supported, especially if the
goal is to use this as a learning tool. (i.e. Reducing the friction between
switching modes.)

If the goal were to use this as a primary input method, redefining all
single-key chords would be more worth it.

The main issue I can see with truly using both at once is learning bad
steno habits, e.g. using chords with the wrong fingers because you decide
to start using them midsentence while typing.

**Comment by user202729** [comment]
Prototype: https://github.com/user202729/plover-half-keyboard

Currently it still detect incorrectly a lot of chord/keys pair (you need to hold the chord for a little. But not too long), but it isn't very easy to come up with a solid heuristic to distinguish between them;

besides is the current capturing code even fast enough? (I feel that it isn't...)

------

By the way this is a small program to visualize the key press/release time (requires `plotly-latest.min.js
` to be placed in the same folder)


[b.zip](https://github.com/openstenoproject/plover/files/6359309/b.zip)

**Comment by user202729** [comment]
Update (*although I work on a different project recently*): The timing parameters turns out to be somewhat hard to get correct...

Have anyone done something similar before?
